### ISSUE TITLE
Multi-project library and transitive dependencies

### ISSUE THREAD

**Issue by rmangi** [clojurephant/clojurephant#128]
## Expected Behavior

This is similar to https://github.com/clojurephant/clojurephant/issues/109

In a multi-project build I have an application A that depends on a sibling library B. B has a few third party dependencies. When compileClojure runs on  A, B's third party libraries should be on the classpath.

## Current Behavior

Currently, I see B (`franz` in the logs) compile fist, as expected but when A (`galileo`) compiles it throws an exception that it can't find a dependency of `franz` (`org.apache.kafka.common.TopicPartition`)

```

> Task :cljbeat:franz:checkClojure
Warning: protocol #'cb.cljbeat.franz/ManualCommittingFranzMachine is overwriting method init of protocol AutocommittingFranzMachine
Warning: protocol #'cb.cljbeat.franz/ManualCommittingFranzMachine is overwriting method step of protocol AutocommittingFranzMachine

> Task :cljbeat:franz:compileClojure
Warning: protocol #'cb.cljbeat.franz/ManualCommittingFranzMachine is overwriting method init of protocol AutocommittingFranzMachine
Warning: protocol #'cb.cljbeat.franz/ManualCommittingFranzMachine is overwriting method step of protocol AutocommittingFranzMachine

> Task :astronomy:galileo:checkClojure FAILED
ERROR Syntax error compiling at (cb/galileo/pipeline.clj:1:1).
Exception in thread "main" Syntax error compiling at (cb/galileo/pipeline.clj:1:1).
        at clojure.lang.Compiler.load(Compiler.java:7647)
        at clojure.lang.RT.loadResourceScript(RT.java:381)
        at clojure.lang.RT.loadResourceScript(RT.java:372)
        at clojure.lang.RT.load(RT.java:463)
        at clojure.lang.RT.load(RT.java:428)
        at clojure.core$load$fn__6824.invoke(core.clj:6126)
        at clojure.core$load.invokeStatic(core.clj:6125)
        at clojure.core$load.doInvoke(core.clj:6109)
        at clojure.lang.RestFn.invoke(RestFn.java:408)
        at dev.clojurephant.tools.clojure_loader$_main$fn__174.invoke(clojure_loader.clj:30)
        at dev.clojurephant.tools.clojure_loader$_main.invokeStatic(clojure_loader.clj:23)
        at dev.clojurephant.tools.clojure_loader$_main.doInvoke(clojure_loader.clj:19)
        at clojure.lang.RestFn.invoke(RestFn.java:397)
        at clojure.lang.AFn.applyToHelper(AFn.java:152)
        at clojure.lang.RestFn.applyTo(RestFn.java:132)
        at clojure.lang.Var.applyTo(Var.java:705)
        at clojure.core$apply.invokeStatic(core.clj:665)
        at clojure.main$main_opt.invokeStatic(main.clj:491)
        at clojure.main$main_opt.invoke(main.clj:487)
        at clojure.main$main.invokeStatic(main.clj:598)
        at clojure.main$main.doInvoke(main.clj:561)
        at clojure.lang.RestFn.applyTo(RestFn.java:137)
        at clojure.lang.Var.applyTo(Var.java:705)
        at clojure.main.main(main.java:37)
Caused by: java.lang.ExceptionInInitializerError
        at java.lang.Class.forName0(Native Method)
        at java.lang.Class.forName(Class.java:348)
        at clojure.lang.RT.classForName(RT.java:2207)
        at clojure.lang.RT.classForName(RT.java:2216)
        at clojure.lang.RT.loadClassForName(RT.java:2235)
        at clojure.lang.RT.load(RT.java:453)
        at clojure.lang.RT.load(RT.java:428)
        at clojure.core$load$fn__6824.invoke(core.clj:6126)
        at clojure.core$load.invokeStatic(core.clj:6125)
        at clojure.core$load.doInvoke(core.clj:6109)
        at clojure.lang.RestFn.invoke(RestFn.java:408)
        at clojure.core$load_one.invokeStatic(core.clj:5908)
        at clojure.core$load_one.invoke(core.clj:5903)
        at clojure.core$load_lib$fn__6765.invoke(core.clj:5948)
        at clojure.core$load_lib.invokeStatic(core.clj:5947)
        at clojure.core$load_lib.doInvoke(core.clj:5928)
        at clojure.lang.RestFn.applyTo(RestFn.java:142)
        at clojure.core$apply.invokeStatic(core.clj:667)
        at clojure.core$load_libs.invokeStatic(core.clj:5985)
        at clojure.core$load_libs.doInvoke(core.clj:5969)
        at clojure.lang.RestFn.applyTo(RestFn.java:137)
        at clojure.core$apply.invokeStatic(core.clj:667)
        at clojure.core$require.invokeStatic(core.clj:6007)
        at clojure.core$require.doInvoke(core.clj:6007)
        at clojure.lang.RestFn.invoke(RestFn.java:512)
        at cb.cljbeat.franz$loading__6706__auto____176.invoke(franz.clj:1)
        at cb.cljbeat.franz__init.load(Unknown Source)
        at cb.cljbeat.franz__init.<clinit>(Unknown Source)
        at java.lang.Class.forName0(Native Method)
        at java.lang.Class.forName(Class.java:348)
        at clojure.lang.RT.classForName(RT.java:2207)
        at clojure.lang.RT.classForName(RT.java:2216)
        at clojure.lang.RT.loadClassForName(RT.java:2235)
        at clojure.lang.RT.load(RT.java:453)
        at clojure.lang.RT.load(RT.java:428)
        at clojure.core$load$fn__6824.invoke(core.clj:6126)
        at clojure.core$load.invokeStatic(core.clj:6125)
        at clojure.core$load.doInvoke(core.clj:6109)
        at clojure.lang.RestFn.invoke(RestFn.java:408)
        at clojure.core$load_one.invokeStatic(core.clj:5908)
        at clojure.core$load_one.invoke(core.clj:5903)
        at clojure.core$load_lib$fn__6765.invoke(core.clj:5948)
        at clojure.core$load_lib.invokeStatic(core.clj:5947)
        at clojure.core$load_lib.doInvoke(core.clj:5928)
        at clojure.lang.RestFn.applyTo(RestFn.java:142)
        at clojure.core$apply.invokeStatic(core.clj:667)
        at clojure.core$load_libs.invokeStatic(core.clj:5985)
        at clojure.core$load_libs.doInvoke(core.clj:5969)
        at clojure.lang.RestFn.applyTo(RestFn.java:137)
        at clojure.core$apply.invokeStatic(core.clj:667)
        at clojure.core$require.invokeStatic(core.clj:6007)
        at clojure.core$require.doInvoke(core.clj:6007)
        at clojure.lang.RestFn.invoke(RestFn.java:805)
        at cb.galileo.pipeline$eval188$loading__6706__auto____189.invoke(pipeline.clj:1)
        at cb.galileo.pipeline$eval188.invokeStatic(pipeline.clj:1)
        at cb.galileo.pipeline$eval188.invoke(pipeline.clj:1)
        at clojure.lang.Compiler.eval(Compiler.java:7176)
        at clojure.lang.Compiler.eval(Compiler.java:7165)
        at clojure.lang.Compiler.load(Compiler.java:7635)
        ... 23 more
Caused by: java.lang.ClassNotFoundException: org.apache.kafka.common.TopicPartition
        at java.net.URLClassLoader.findClass(URLClassLoader.java:381)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:424)
        at sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:335)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
        at java.lang.Class.forName0(Native Method)
        at java.lang.Class.forName(Class.java:348)
        at clojure.lang.RT.classForName(RT.java:2207)
        at clojure.lang.RT.classForNameNonLoading(RT.java:2220)
        at cb.cljbeat.franz.consumer$loading__6706__auto____178.invoke(consumer.clj:1)
        at cb.cljbeat.franz.consumer__init.load(Unknown Source)
        at cb.cljbeat.franz.consumer__init.<clinit>(Unknown Source)
        ... 82 more
```

I tried doing the same but using the java-library plugin and setting the dependencies to `api` (they are currently `implementation`) but that failed to put the code for my library on the application classpath at all. 

The deps for the application looks like this

```
dependencies {
  implementation commonDeps.cljcore
  implementation commonDeps.logging
  implementation commonDeps.component
  implementation commonDeps.cbcore
  implementation commonDeps.core_cache
  implementation 'org.flatland:ordered:1.5.7'
  implementation project(':cljbeat:franz')
  implementation commonDeps.msgpack
  implementation cbDeps.pingdecoder
  devImplementation commonDeps.devtools
  testImplementation 'junit:junit:4.12'
}
```

and for the library 

```
dependencies {
  implementation commonDeps.cljcore
  implementation 'clojurewerkz:propertied:1.3.0'
  implementation 'org.apache.kafka:kafka-clients:0.10.2.0'
  devImplementation commonDeps.devtools
  testImplementation 'junit:junit:4.12'
}
```

Thanks! This is a great project. Not having a reliable way to manage multi-project builds in clojure is a major issue for us as our code base grows. 

- **clojurephant version:** 0.5.0-alpha.5
- **Gradle version:** 5.4.1
- **Java version:** 1.8
- **OS version:** OSX

**Comment by ajoberstar** [comment]
Interesting, I'll have to make a test to reproduce this and see where it stems from.

**Comment by ajoberstar** [comment]
Sorry, I haven't been devoting much time to open source stuff recently. If you are interested in helping that would much appreciated. If you need any help getting started please let me know.

**Comment by ajoberstar** [comment]
Did some research on this one today, but it seems a little trickier than I had hoped. I think Gradle will require you to use the `api` configuration instead of `implementation`, but... there are some other config issues that need to be dealt with to confirm and support that.

**Comment by pbzdyl** [comment]
Hi @ajoberstar, it seems my project is impacted by this issue too. Are you planning to take a look at this in the near future? Or do you have any pointers to how it could be fixed so I could try to work on this?

**Comment by ajoberstar** [comment]
I have some ideas on this one, and hoping to work on it soon. I think it's going to involve adding a new variant for aot and simplifying some of the other code as a result.

It's getting intertwined with some other work I'm doing, so I'll see if I can split it out.

**Comment by ajoberstar** [comment]
So, I think this just requires you to put any of your dependencies as `api` dependencies since that's how Gradle (particularly with the `java-library` plugin) exports to downstream projects. This is because they're used to the Java model of being able to just export the API layer and compile against that. Since Clojure has to load it all, it's going to seem a little strange, but it should do the trick.

If there are further issues, I think they'd be solved by the fix for #134.

**Comment by pbzdyl** [comment]
Thank you @ajoberstar!
